src/Features/Core/Portable/Ambiguity/AbstractAmbiguousTypeCodeFixProvider.cs

```diff
@@ -0,0 +1,103 @@
+                var typeName = GetAliasFromDiagnosticNode(syntaxFacts, diagnosticNode);
+                foreach (var symbol in symbolInfo.CandidateSymbols)
+                {
+                    var aliasDirective = GetAliasDirective(typeName, symbol);
+                    var codeActionPreviewText = await GetTextPreviewOfChangeAsync(aliasDirective,
+                                                                                  document.Project.Solution.Workspace,
+                                                                                  optionSet,
+                                                                                  cancellationToken).ConfigureAwait(false);
+                    var newRoot = addImportService.AddImport(compilation, root, diagnosticNode, aliasDirective, placeSystemNamespaceFirst);
+                    var codeAction = new MyCodeAction(codeActionPreviewText, c => Task.FromResult(document.WithSyntaxRoot(newRoot)));
+                    context.RegisterCodeFix(codeAction, context.Diagnostics.First());
+                }
+            }
+        }
+
+        private static async Task<string> GetTextPreviewOfChangeAsync(SyntaxNode newNode, Workspace workspace, OptionSet optionSet, CancellationToken cancellationToken)
+        {
+            var formattedNode = await Formatter.FormatAsync(newNode, workspace, optionSet, cancellationToken).ConfigureAwait(false);
+            var formattedText = formattedNode.ToFullString();
+            return string.Format(FeaturesResources.Alias_ambiguous_type_0, formattedText);
+        }
+
+        private static string GetAliasFromDiagnosticNode(ISyntaxFactsService syntaxFacts, SyntaxNode diagnosticNode)
+        {
+            // The content of the node is a good candidate for the alias
+            // For attributes VB requires that the alias ends with 'Attribute' while C# is fine with or without the suffix.
+            var nodeText = diagnosticNode.ToString();
+            if (syntaxFacts.IsAttribute(diagnosticNode.Parent))
+            {
+                if (!nodeText.EndsWith("Attribute"))
+                {
+                    nodeText += "Attribute";
+                }
+            }
+
+            return nodeText;
+        }
+
+        private static bool SymbolCandidatesContainsSupportedSymbols(SymbolInfo symbolInfo)
+            => symbolInfo.CandidateReason == CandidateReason.Ambiguous &&
+               // Arity: Aliases can only name closed constructed types.
+               // Aliasing as a closed constructed type is possible but would require to remove the type arguments from the diagnosed node.
+               // It is unlikely that the user wants that and so generic types are not supported.
+               symbolInfo.CandidateSymbols.All(symbol => symbol.IsKind(SymbolKind.NamedType) &&
+                                                         symbol.GetArity() == 0);
```

Comment:
ah, you check that it's a named type. good.  

probably good to check that its containing symbol is a namespace as well.

Reply:
> probably good to check that its containing symbol is a namespace as well.

Nested classes can also be ambigues. The corresponding test is [here](
https://github.com/dotnet/roslyn/pull/24022/files#diff-871cc5d400cdd54881c965ca53a634eeR208). The test case is copied from NameCollisionTests.cs

Reply:
Gotcha.  Are there any issues if there are generics in the containing type (or an intermediary generic type)?  might be good to have tests.

Reply:
I think the ambiguity for nested types is only possible with generic classes in combination with using static. So containing generic types are already tested. The test is called TestAmbiguousNestedClass in line 208 in AmbiguousTypeTests.cs (and can't be linked because the test file is to big for github).